### PR TITLE
HDDS-11413. PipelineManagerImpl lock optimization reduces AllocateBlock latency

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

PipelineManagerImpl lock optimization reduces AllocateBlock latency

The following are the issues discovered in our production environment:
When applying for a block in the Ratis replica, there may be jitter in the P99 latency due to obtaining a read lock
image

The main parts that hold write locks:

```
org.apache.hadoop.hdds.scm.pipeline.WritableECContainerProvider#allocateContainer
```

image

EC pipelines are frequently created and closed because there is only one container in each EC pipeline. When the container is full, the creation of new ones will be closed. When creating a pipeline, a write lock at the pipeline manager level will be held, while when applying for a Ratis block, a read lock at the pipeline manager level will be held.

The purpose of this optimization is to unlock. When creating and closing an EC pipeline, only the write lock at the EC pipeline manager level will be held, while when applying for a ratis replica block, only the read lock of the ratis pipeline manager will be held. The two locks will not affect each other.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-11413

### CODE REVIEW

**Comment by kerneltime** [comment]
Thank you for making this improvement. Can you please post the measurements done before and after for this performance improvement?

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineManager.java [hunk 1]
```diff
@@ -99,29 +100,34 @@ public static ReconPipelineManager newReconPipelineManager(
...
         }
+      } finally {
+        releaseWriteLock(replicationConfig);
       }
-      removeInvalidPipelines(pipelinesFromScm);
-    } finally {
-      releaseWriteLock();
+
     }
+    removeInvalidPipelines(pipelinesFromScm);
```
Comment (xichen01) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineManager.java:129]: Should we call it inside the write lock?

**Review by xichen01 — COMMENTED** [review]
Thanks for your patch, please see the comments inlined.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineManager.java (hunk 1 above)
Comment (guohao-rosicky) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/scm/ReconPipelineManager.java]: @xichen01 Thanks for the review. Have been modified.

**Comment by adoroszlai** [comment]
@guohao-rosicky can you please share performance improvement numbers as requested in [this comment](https://github.com/apache/ozone/pull/7160#issuecomment-2338581312):

> Can you please post the measurements done before and after for this performance improvement?

**Comment by jojochuang** [comment]
Looks good.

Also looks like we don't have metrics in SCM for allocateBlock. That would make finding p99 latency problems easier.

**Comment by adoroszlai** [comment]
> Looks good.

@jojochuang are you comfortable merging this?

**Comment by jojochuang** [comment]
Merging this one. I'll open another jira for the SCM allocateBlock metric.

**Comment by swamirishi** [comment]
@guohao-rosicky @jojochuang currently CI is failing on TestBlockManager test failure for 
org.apache.hadoop.hdds.scm.block.TestBlockManager#testAllocateBlockInParallel
org.apache.hadoop.hdds.scm.block.TestBlockManager#testBlockAllocationWithNoAvailablePipelines
org.apache.hadoop.hdds.scm.block.TestBlockManager#testAllocateBlockSucInSafeMode.
Can we revert this patch and triage the test case failures?

**Pull request merged**